client/js/analysis/gdpr.js

```diff
@@ -67,14 +121,95 @@ export async function getStreamingData(data) {
           Promise
             .all(contentPromises)
             .then((contentWithIndices) => {
-              const content = contentWithIndices
-                .map((entry) => entry.data)
-                .reduce((dataA, dataB) => dataA.concat(dataB), []);
-              resolve(content);
+              resolve(
+                contentWithIndices
+                  .map((entry) => entry.data)
+                  .reduce((dataA, dataB) => dataA.concat(dataB), [])
+                  .map((entry) => {
+                    // dates are strings in JSON, convert to JS date
+                    let endTime = new Date(entry.endTime);
+                    // move according to UTC offset
+                    endTime = new Date(+endTime - endTime.getTimezoneOffset() * 60000);
+                    entry.endTime = endTime;
+                    return entry;
+                  }),
+              );
             });
         });
       },
       (error) => reject(error),
     );
   });
 }
+
+/** @typedef {Map<string, Map<string, number>>} ArtistTrackTimeMap */
+/** @typedef {{
+ * start: Date,
+ * totals: ArtistTrackTimeMap
+ * }} ArtistTrackTimeWindow */
+
+/**
+ * Calculates aggregates from a list of GDPR records.
+ *
+ * @param {GDPRRecord[]} data The list of GDPR records to aggregate.
+ * @returns {{totals: ArtistTrackTimeMap, windows: ArtistTrackTimeWindow[]}} An
+ * object with two properties: `totals` and `windows`. `totals` is a map whose
+ * keys are artists and whose values are maps from track names to duration
+ * listened. `windows` is an array of window objects. Each window object has a
+ * `start` representing the start of that window, and a `totals`, representing
+ * the durations listened to each song during that window.
+ */
+export function collateStreamingData(data) {
+  let index = 0;
+  let currentRecord = data[0];
+
+  const windowSize = 86400000;
+  // round down to the start of the day of the first record
+  let windowStart =
+    +currentRecord.endTime - (+currentRecord.endTime % windowSize);
+
+  const windows = [];
+
+  // a map of maps: key is artist name, value is a map where key
+  // is track name and value is time listened to that song in total
+  /** @type {ArtistTrackTimeMap} */
+  const totals = new Map();
+
+  // each window is a map of maps: key is artist name, value is a map where key
+  // is track name and value is time listened to that song during this window
+  /** @type {ArtistTrackTimeMap} */
+  let window = new Map();
+
+  while (index < data.length) {
+    while (currentRecord && +currentRecord.endTime < windowStart + windowSize) {
+      const {artistName, trackName, msPlayed} = currentRecord;
+
+      const windowArtistTracks = getOrDefault(
+        window, artistName, () => new Map(),
+      );
```

Comment:
So basically if there's key is undefined you make a new map. I don't know if you plan to have different functions inside, and it would make sense if you plan to use it later. Otherwise I suggest just making the new map in the actual function from a reader's point of view.

Reply:
I second ^^